design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,429 @@
+overall. Today provider-aws, at idle, must start almost 900 controllers. One for
+each supported kind of managed resource. Each controller spins up a few
+goroutines and subscribes to API server updates for the type of MR it
+reconciles. By breaking up providers significantly fewer controllers would be
+running, using resources at idle.
+
+We expect significantly more compute resources to be consumed by Terraform
+provider processes because there will need to be more of them. It’s safe to
+assume each deployment of provider-aws becomes ~7 smaller deployments on
+average. For a single ProviderConfig, each of these smaller deployments would
+incur the ~300MB of memory required to run a Terraform provider process. This
+doubles as we increase to two ProviderConfigs, etc.
+
+The increased compute resources consumed by needing to run Terraform providers
+is a serious concern. I’m not considering it a blocker for this proposal because
+we feel there is a lot of room for improvement in the medium to long term. In
+addition to the current steps being investigated to improve performance, we
+could consider:
+
+* Replacing the most widely-used MR controllers with native implementations that
+  use dramatically fewer resources.
+* Cutting out Terraform CLI invocations by talking directly to the Terraform
+  provider process.
+* Cutting out both Terraform CLI invocations and Terraform provider processes by
+  importing the Terraform provider code directly. Terraform makes this hard to
+  do, but it’s possible. It appears to be what Config Connector does.
+
+## Alternatives Considered
+
+I considered the following alternatives before making this proposal.
+
+### CRD Filtering
+
+The gist of this proposal is to update the existing Provider.pkg.crossplane.io
+API to include a configurable allow list of enabled CRDs. As far as I can tell
+this is the only feasible alternative to breaking up the largest Providers.
+
+If we took this approach I would propose that:
+
+* We teach the Package Manager that only some Providers support filtering. Such
+  Providers would have something like `pkg.crossplane.io/filtering: enabled` in
+  their `crossplane.yaml`. This is necessary to avoid filtering being silently
+  ineffectual when not supported by a Provider.
+* We add an optional filter list to the Provider.pkg type. This would be
+  replicated to the ProviderRevision.pkg type (that actually installs/owns a
+  Provider’s CRDs).
+* If you installed a filtering-enabled Provider into an empty control plane and
+  didn’t specify what types to enable, no types would be enabled.
+* If you installed a filtering-enabled Provider into a control plane the Package
+  Manager would automatically detect what types were in use and enable them
+  automatically.
+* If you installed a filtering-enabled Provider as a dependency of a
+  Configuration the Package Manager would scan its Compositions for types and
+  enable them automatically.
```

Comment:
We know today that installing a single, sufficiently large provider will degrade some control planes, right? For example we know installing a recent version of `upbound/provider-aws` on a new GKE control plane causes it to become unresponsive. In my mind this makes installing all CRDs by default a non-starter.

Reply:
I assume someone would not install a provider in a prod environment at first, so I agree it's not the best UX "breaking" the cluster at first, but having it clearly documented and maybe explicitly allowing all groups in the Provider's installation manifests to make it even more clear that a lot of CRDs are going to be installed, could make it acceptable IMO.

Reply:
Agree that it is not an ideal solution, but given that there is no perfect solution, I would like us to reevaluate different alternatives here.

What about something in between? Instead of installing all or none by default, can we install a default set of CRDs defined by the providers themselves? I imagine each provider defines a set of APIs enabled by default in their `crossplane.yaml`'s, and it would be used to _initialize_ the `enabledAPIs` in `provider.pkg` by package manager during installation. For upgrades, the package manager will also consider existing MRs and extend the list of enabled APIs. 

I mean; 

- Installing a provider on an empty control plane as today: Package manager checks if `enabledAPIs` configured, if not, it will initialize with ~100 CRDs defined by the provider package. 
- Upgrading provider aws: Package manager checks the existing CRs and merges them with default APIs to initialize `enabledAPIs`.

This will be more like a defaulting mechanism, and the package manager will not interfere with `enabledAPIs`, if configured already.

I believe solutions like this were considered as hairy in the past. Still, given there is no ideal solution, I would like us to investigate "how hairy it is" and whether we can find a relatively complex but working solution.

Reply:
> What about something in between?

I think this shakes out as about the same as defaulting to nothing, right? The key point is that whatever set the Provider defaults to might not be what you're running when you upgrade, so we have to build the "magic" that figures out what you _are_ using and implicitly enables it.

Doable, but it's more magic added to the system.